collector/arp_linux.go

```diff
@@ -69,13 +75,65 @@ func getTotalArpEntries(deviceEntries []procfs.ARPEntry) map[string]uint32 {
 	return entries
 }
 
-func (c *arpCollector) Update(ch chan<- prometheus.Metric) error {
-	entries, err := c.fs.GatherARPEntries()
+func getTotalArpEntriesRTNL() (map[string]uint32, error) {
+	conn, err := rtnetlink.Dial(nil)
+	if err != nil {
+		return nil, err
+	}
+	defer conn.Close()
+
+	neighbors, err := conn.Neigh.List()
 	if err != nil {
-		return fmt.Errorf("could not get ARP entries: %w", err)
+		return nil, err
+	}
+
+	ifIndexEntries := make(map[uint32]uint32)
+
+	for _, n := range neighbors {
+		// Neighbors will also contain IPv6 neighbors, but since this is purely an ARP collector,
+		// restrict to AF_INET. Also skip entries which have state NUD_NOARP to conform to output
+		// of /proc/net/arp.
+		if n.Family == unix.AF_INET && n.State&unix.NUD_NOARP == 0 {
+			ifIndexEntries[n.Index]++
+		}
 	}
 
-	enumeratedEntry := getTotalArpEntries(entries)
+	enumEntries := make(map[string]uint32)
+
+	// Convert interface indexes to names.
+	for ifIndex, entryCount := range ifIndexEntries {
+		iface, err := net.InterfaceByIndex(int(ifIndex))
+		if err != nil {
+			if errors.Unwrap(err).Error() == "no such network interface" {
+				continue
+			}
+			return nil, err
+		}
+
+		enumEntries[iface.Name] = entryCount
+	}
+
+	return enumEntries, nil
+}
+
+func (c *arpCollector) Update(ch chan<- prometheus.Metric) error {
+	var enumeratedEntry map[string]uint32
+
+	if *arpNetlink {
+		var err error
+
+		enumeratedEntry, err = getTotalArpEntriesRTNL()
+		if err != nil {
+			return fmt.Errorf("could not get ARP entries: %w", err)
+		}
+	} else {
+		entries, err := c.fs.GatherARPEntries()
```

Comment:
minor nit for consistency: directly assign to enumeratedEntry or to the same you do here in the clause above

Reply:
Ah I see, yeah missed that you need to call getTotalArpEntries. Never mind then!